File: 2014년문제답안/Problem03/house_parallel.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <sys/time.h>
#include "mpi.h"
#define GRANULARITY 5		// it should be less than nx
#define MASTER 0



int nprocs, myrank;


struct job {
	bool working;			// true:working, false:quit
	int site[GRANULARITY];
};


struct job_msg {
	int solutions_found;
	int origin;
};



double time_diff(struct timeval t1, struct timeval t2) {
	double sec1, sec2;

	sec1 = t1.tv_sec + t1.tv_usec*1e-6;
	sec2 = t2.tv_sec + t2.tv_usec*1e-6;

	return sec2-sec1;
}



int master_build_house(int nx, int column, int site[]) {
   	int num_sol = 0;
	bool is_sol;
	int i, j;

	for (i=0; i<nx; i++) {
   		// Try to build a house in each line of column
	   	site[column] = i;

	   	// Check if this placement is still a solution
		is_sol = true;
		for (j=column-1; j>=0; j--) {
		   	if ((site[column] == site[j])               ||
				(site[column] == site[j] - (column-j))  ||
				(site[column] == site[j] + (column-j))) {
			   	is_sol = false;
			   	break;
		   	}
	   	}

	   	if (is_sol) {
		  	if (column == GRANULARITY-1) {
				// If this is the last level (granularity of the job),
				// send a next job to a worker
				num_sol += send_job_worker(nx, site);
		   	} 
			else {
			   	// The placement is not complete.
				// Try to place the house on the next column
				num_sol += master_build_house(nx, column+1, site);
		   	}
	   	}
   	}

   	return num_sol;
}



int worker_build_house(int nx, int column, int sub_site[]) {
   	int num_sol = 0;
	bool is_sol;
	int i, j;

   	// Try to build a house in each line of column
	for (i=0; i<nx; i++) {
	   	sub_site[column] = i;

	   	// Check if this placement is still a solution
		is_sol = true;
		for (j=column-1; j>=0; j--) {
		   	if ((sub_site[column] == sub_site[j])               ||
				(sub_site[column] == sub_site[j] - (column-j))  ||
				(sub_site[column] == sub_site[j] + (column-j))) {
			   	is_sol = false;
			   	break;
		   	}
	   	}

	   	if (is_sol) { 
		  	if (column == nx-1) {
			   	// If this is the last column, printout the solution
				num_sol += 1;

		   	} else {
			   	// The placement is not complete.
				// Try to place the house on the next column
				num_sol += worker_build_house(nx, column+1, sub_site);
		   	}
	   	}
   	}

   	return num_sol;
}



int send_job_worker(int nx, int site[]) {
	int num_sol = 0;		// The number of solutions found meanwhile
	int i;
	struct job todo;
	struct job_msg msg;
	
	// Set the job
	todo.working = true;

	for (i=0; i<GRANULARITY; i++)
		todo.site[i] = site[i];

	// Recieve the last result from a worker
	MPI_Recv(&msg, sizeof(msg), MPI_BYTE, MPI_ANY_SOURCE, MPI_ANY_TAG, 
			 MPI_COMM_WORLD, MPI_STATUS_IGNORE);

	num_sol = msg.solutions_found;

	// Send the new job to the worker
	MPI_Send(&todo, sizeof(todo), MPI_BYTE, msg.origin, 0, MPI_COMM_WORLD);

	return num_sol;
}



int wait_remaining_results() {
	// Wait for remaining results, sending a quit whenever a new result arrives
	
	int num_sol = 0;
	int n_workers = nprocs-1;
	struct job todo;
	struct job_msg msg;

	todo.working = false;

	while (n_workers > 0) {
		// Receive a message from a worker
		MPI_Recv(&msg, sizeof(msg), MPI_BYTE, MPI_ANY_SOURCE, MPI_ANY_TAG, 
				 MPI_COMM_WORLD, MPI_STATUS_IGNORE);
		num_sol += msg.solutions_found;

		MPI_Send(&todo, sizeof(todo), MPI_BYTE, msg.origin, 0, MPI_COMM_WORLD);

		n_workers -= 1;
	}

	return num_sol;
	
}



void worker(int nx) {
	// There is a default message which lets a worker request a 
	// job reporting the number of solutions found in the last iteration
	
	int num_sol;
	struct job_msg msg;

    msg.origin = myrank;
	msg.solutions_found = 0;

	// Request initial job
	MPI_Send(&msg, sizeof(msg), MPI_BYTE, MASTER, 0, MPI_COMM_WORLD);

	while (true) {
		// Wait for a job or a quit message
	    struct job todo;
		MPI_Recv(&todo, sizeof(todo), MPI_BYTE, MPI_ANY_SOURCE, 
				 MPI_ANY_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

		if (todo.working == false)
		    break;

		num_sol = worker_build_house(nx, GRANULARITY, todo.site);

		// Ask for more work
		msg.origin = myrank;
		msg.solutions_found = num_sol;
		MPI_Send(&msg, sizeof(msg), MPI_BYTE, MASTER, 0, MPI_COMM_WORLD);
	}
}



void main(int argc, char* argv[]) {
	int num_sol;
	int nx=14;
	int *site;
	struct timeval start, end;

	MPI_Init(&argc, &argv);
	MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
	MPI_Comm_size(MPI_COMM_WORLD, &nprocs);


	if (myrank == MASTER) {
		printf("nx=%d\n", nx);
		printf("nprocs=%d\n", nprocs);
		site = (int*)malloc(nx*sizeof(int));

		gettimeofday(&start, NULL);
		num_sol = master_build_house(nx, 0, site);
		num_sol += wait_remaining_results();
		gettimeofday(&end, NULL);

		printf("num_sol=%d\n", num_sol);
		printf("elapsed time=%.3lf sec\n", time_diff(start,end));
	}
	else {
		worker(nx);
	}

	MPI_Finalize();
}

File: 2014년문제답안/Problem03/house_parallel.f90
PROGRAM houses
  USE MPI
  IMPLICIT NONE
  INTEGER, PARAMETER :: nx=17
  INTEGER, PARAMETER :: GRANULARITY=5   ! it should be less than nx

  TYPE :: JOB
      SEQUENCE
      LOGICAL :: working        ! true: working, false: quit
      INTEGER, DIMENSION(GRANULARITY) :: site
  END TYPE

  TYPE :: JOB_MSG
      SEQUENCE
      INTEGER :: solutions_found
      INTEGER :: origin
  END TYPE

  INTEGER, DIMENSION(nx) :: site
  INTEGER :: total_num_sol
  CHARACTER(10) :: date, time, zone     ! measure time
  INTEGER, DIMENSION(8) :: t1, t2
  REAL :: elapsed
  INTEGER :: nprocs, myrank, ierr, stat(MPI_STATUS_SIZE)      ! MPI variables


  CALL MPI_INIT(ierr)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nprocs, ierr)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, myrank, ierr)


  IF (myrank == 0) THEN
    PRINT *, 'nx=', nx
    PRINT *, 'nprocs=', nprocs

    CALL DATE_AND_TIME(date, time, zone, t1)
    total_num_sol = master_build_house(1)
    total_num_sol = total_num_sol + wait_remaining_results()
    CALL DATE_AND_TIME(date, time, zone, t2)
    elapsed = REAL(t2(5)-t1(5))*3600 + REAL(t2(6)-t1(6))*60 + &
              REAL(t2(7)-t1(7))      + REAL(t2(8)-t1(8))*1e-3   ! sec

    PRINT *, 'num_sol=', total_num_sol
    PRINT *, 'elapsed_time=', elapsed, 'sec'

  ELSE
    CALL worker()
  END IF

  CALL MPI_FINALIZE(ierr)


CONTAINS
  RECURSIVE FUNCTION master_build_house(column) RESULT(num_sol)
    IMPLICIT NONE
    INTEGER, INTENT(IN) :: column
    LOGICAL :: is_sol
    INTEGER :: num_sol, i, j
    
    num_sol = 0

    DO i=1,nx
      ! Try to build a house in each line of column
      site(column) = i

      ! Check if this placement is still a solution
      is_sol = .TRUE.
      DO j=column-1,1,-1
        IF (site(column) == site(j) .OR. &
            site(column) == site(j)-(column-j) .OR. &
            site(column) == site(j)+(column-j)) THEN
          is_sol = .FALSE.
          EXIT
        END IF
      END DO

      IF (is_sol) THEN
        IF (column == GRANULARITY) THEN
          ! If this is the last level (granularity of the job),
          ! send a next job to a worker
          num_sol = num_sol + send_job_worker()
        ELSE   
          ! The placement is not complete.
          ! Try to place the house on the next column
          num_sol = num_sol + master_build_house(column+1)
        END IF
      END IF
    END DO
  END FUNCTION



  RECURSIVE FUNCTION worker_build_house(column, sub_site) RESULT(num_sol)
    IMPLICIT NONE
    INTEGER, INTENT(IN) :: column
    INTEGER, DIMENSION(GRANULARITY), INTENT(OUT) :: sub_site
    LOGICAL :: is_sol
    INTEGER :: num_sol, i, j

    num_sol = 0

    DO i=1,nx
      ! Try to build a house in each line of column
      sub_site(column) = i

      ! Check if this placement is still a solution
      is_sol = .TRUE.
      DO j=column-1,1,-1
        IF (sub_site(column) == sub_site(j) .OR. &
            sub_site(column) == sub_site(j)-(column-j) .OR. &
            sub_site(column) == sub_site(j)+(column-j)) THEN
          is_sol = .FALSE.
          EXIT
        END IF
      END DO

      IF (is_sol) THEN
        IF (column == nx) THEN
          ! If this is the last column, printout the solution
          num_sol = num_sol + 1
        ELSE   
          ! The placement is not complete.
          ! Try to place the house on the next column
          num_sol = num_sol + worker_build_house(column+1, sub_site)
        END IF
      END IF
    END DO
  END FUNCTION



  FUNCTION send_job_worker() RESULT(num_sol)
    IMPLICIT NONE
    LOGICAL :: is_sol
    INTEGER :: num_sol, i, j
    TYPE(JOB) :: todo
    TYPE(JOB_MSG) :: msg

    num_sol = 0     ! The number of solutions found meanwhile

    ! Set the job
    todo%working = .TRUE.

    DO i=1,GRANULARITY
      todo%site(i) = site(i)
    END DO

    ! Recieve the last result from a worker
    CALL MPI_RECV(msg, SIZEOF(msg), MPI_BYTE, MPI_ANY_SOURCE, 0, MPI_COMM_WORLD, stat, ierr)

    num_sol = msg%solutions_found

    ! Send the new job to the worker
    CALL MPI_SEND(todo, SIZEOF(todo), MPI_BYTE, msg%origin, 1, MPI_COMM_WORLD, ierr)
  END FUNCTION



  FUNCTION wait_remaining_results() RESULT(num_sol)
    ! Wait for remaining results, sending a quit whenever a new result arrives
    IMPLICIT NONE
    INTEGER :: num_sol, n_workers
    TYPE(JOB) :: todo
    TYPE(JOB_MSG) :: msg

    num_sol = 0
    n_workers = nprocs-1

    todo%working = .FALSE.

    DO WHILE (n_workers > 0)
      ! Receive a message from a worker
      CALL MPI_RECV(msg, SIZEOF(msg), MPI_BYTE, MPI_ANY_SOURCE, 0, MPI_COMM_WORLD, stat, ierr)
      num_sol = num_sol + msg%solutions_found

      CALL MPI_SEND(todo, SIZEOF(todo), MPI_BYTE, msg%origin, 1, MPI_COMM_WORLD, ierr)

      n_workers = n_workers - 1;
    END DO
  END FUNCTION



  SUBROUTINE worker()
    ! There is a default message which lets a worker request a 
    ! job reporting the number of solutions found in the last iteration
    IMPLICIT NONE
    INTEGER :: num_sol, i
    TYPE(JOB) :: todo
    TYPE(JOB_MSG) :: msg

    msg%solutions_found = 0
    msg%origin = myrank

    ! Request initial job
    CALL MPI_SEND(msg, SIZEOF(msg), MPI_BYTE, 0, 0, MPI_COMM_WORLD, ierr)

    DO WHILE (.TRUE.)
      ! Wait for a job or a quit message
      CALL MPI_RECV(todo, SIZEOF(todo), MPI_BYTE, 0, 1, MPI_COMM_WORLD, stat, ierr)

      IF (.NOT. todo%working) EXIT

      num_sol = worker_build_house(GRANULARITY+1, todo%site)

      ! Ask for more work
      msg%solutions_found = num_sol
      msg%origin = myrank
      CALL MPI_SEND(msg, SIZEOF(msg), MPI_BYTE, 0, 0, MPI_COMM_WORLD, ierr)
    END DO
  END SUBROUTINE
END PROGRAM
